instructionAPI/src/Instruction.C

```diff
@@ -97,26 +97,16 @@ namespace Dyninst { namespace InstructionAPI {
 
   DYNINST_EXPORT Instruction::Instruction(Operation what, size_t size, const unsigned char* raw,
                                           Dyninst::Architecture arch)
-      : m_InsnOp(what), m_Valid(is_valid_mnemonic(arch, what.getID())), arch_decoded_from(arch),
-        formatter(&ArchSpecificFormatter::getFormatter(arch)) {
+      : m_InsnOp(what), m_Valid(is_valid_mnemonic(arch, what.getID())), m_size{static_cast<decltype(m_size)>(size)},
+        arch_decoded_from(arch), formatter(&ArchSpecificFormatter::getFormatter(arch)) {
     copyRaw(size, raw);
   }
 
   void Instruction::copyRaw(size_t size, const unsigned char* raw) {
-    if(m_size > sizeof(m_RawInsn.small_insn)) {
-        delete[] m_RawInsn.large_insn; 
-    }
-    m_size = 0;
-    m_RawInsn.small_insn = 0;
-    if(raw) {
-      m_size = size;
-      if(size <= sizeof(m_RawInsn.small_insn)) {
-        memcpy(&m_RawInsn.small_insn, raw, size);
-      } else {
-        m_RawInsn.large_insn = new unsigned char[size];
-        memcpy(m_RawInsn.large_insn, raw, size);
-      }
-    }
+    auto last = std::copy_n(raw, size, m_RawInsn.data());
```

Comment:
Does m_size needs to be updated when copyRaw is called, if you fill the rest with 0?

Reply:
It's set in the constructors, but I forgot to update it when called from `updateSize`. Fixed.

Reply:
> I tried to set up `m_size` as a class invariant since this function will go away once the decoders have a sane interface for constructing an Instruction. I'm still not entirely certain why the AMDGPU decoder needs to add bytes to the opcode mid-decode, but I'll leave that for another day.

AMDGPU instructions can use a 32 bit literal constant, but whether or not it is using the constant can only be determined after you decode its operand.

Reply:
> AMDGPU instructions can use a 32 bit literal constant, but whether or not it is using the constant can only be determined after you decode its operand.

I see. Does that mean that every instruction has to be speculatively decoded?